File: java/src/gov/usgs/hazdevbroker/Consumer.java
package gov.usgs.hazdevbroker;

import java.util.*;
import java.io.FileNotFoundException;
import java.io.UnsupportedEncodingException;
import org.apache.log4j.Logger;

import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.common.PartitionInfo;
import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;

/**
 * A hazdev broker class used to poll data from the hazdev kafka broker cluster.
 *
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class Consumer extends ClientBase {

	/**
	 * Optional configuration string defining the heartbeat directory
	 */
	private static String heartbeatDirectory;

	/**
	 * The kafka consumer client
	 */
	private static org.apache.kafka.clients.consumer.Consumer<String, byte[]> consumer;

	/**
	 * The Heartbeat message processor
	 */
	private static Heartbeat heartbeatProcessor;

	/**
	 * Variable containing time of the last heartbeat.
	 */
	private static Long lastHeartbeatTime;

	/**
	 * A collection of strings contining the topics 
	 */	
	private static Collection<String> topicList;

	/**
	 * Log4J logger for Consumer
	 */
	static Logger logger = Logger.getLogger(Consumer.class);

	/**
	 * The constructor for the Consumer class. Initializes members to default or
	 * null values.
	 */
	public Consumer() {
		// init
		consumer = null;
		heartbeatProcessor = null;
		heartbeatDirectory = null;
		topicList = null;
		CONFIGTYPE_STRING = "ConsumerConfig";

		// init last  heartbeat time to now
		lastHeartbeatTime = (Long) (System.currentTimeMillis() / 1000);	
	}

	/**
	 * The advanced constructor for the Consumer class. Initializes members to
	 * default values and used the provided JSON configuration to configure the
	 * kafka consumer client.
	 *
	 * @param configObject
	 *            - A JSONObject containing the configuration
	 */
	public Consumer(JSONObject configObject) {
		// init
		consumer = null;
		heartbeatProcessor = null;
		heartbeatDirectory = null;
		topicList = null;
		CONFIGTYPE_STRING = "ConsumerConfig";

		// init last  heartbeat time to now
		lastHeartbeatTime = (Long) (System.currentTimeMillis() / 1000);		

		// configuration/setup
		Properties configuration = convertJSONConfigToProp(configObject);
		setup(configuration);
	}

/**
	 * The advanced constructor for the Consumer class. Initializes members to
	 * default values and used the provided JSON configuration to configure the
	 * kafka consumer client.
	 *
	 * @param configObject
	 *            - A JSONObject containing the configuration
	 * @param hbDirectory
	 * 			  - A String containing the directory to write heartbeat messages
	 * Set to null to disable heartbeat messages
	 */
	public Consumer(JSONObject configObject, String hbDirectory) {
		// init
		consumer = null;
		heartbeatProcessor = null;
		heartbeatDirectory = hbDirectory;
		topicList = null;
		CONFIGTYPE_STRING = "ConsumerConfig";

		// init last  heartbeat time to now
		lastHeartbeatTime = (Long) (System.currentTimeMillis() / 1000);		

		// configuration/setup
		Properties configuration = null;
		try {
			configuration = convertJSONConfigToProp(configObject);
		} catch (Exception e) {
			logger.error("Exception converting configuration: " + e.toString());
			throw (e);
		}

		setup(configuration);	
	}

	/**
	 * The advanced constructor for the Consumer class. Initializes members to
	 * default values and used the provided JSON configuration string to
	 * configure the kafka consumer client.
	 *
	 * @param configString
	 *            - A JSON formatted String containing the configuration
	 * @throws org.json.simple.parser.ParseException
	 *             if a json parse exception occurs
	 */
	public Consumer(String configString) throws ParseException {
		// init
		consumer = null;
		heartbeatProcessor = null;
		heartbeatDirectory = null;
		topicList = null;
		CONFIGTYPE_STRING = "ConsumerConfig";

		// init last  heartbeat time to now
		lastHeartbeatTime = (Long) (System.currentTimeMillis() / 1000);		

		// configuration/setup
		Properties configuration = null;
		try {
			configuration = convertJSONStringToProp(configString);
		} catch (ParseException e) {
			logger.error("ParseException converting configuration: " + e.toString());
			throw (e);
		}

		setup(configuration);	
	}

	/**
	 * The advanced constructor for the Consumer class. Initializes members to
	 * default values and used the provided JSON configuration string to
	 * configure the kafka consumer client.
	 *
	 * @param configString
	 *            - A JSON formatted String containing the configuration
	 * @param hbDirectory
	 * 			  - A String containing the directory to write heartbeat messages
	 * @throws org.json.simple.parser.ParseException
	 *             if a json parse exception occurs
	 */
	public Consumer(String configString, String hbDirectory) 
			throws ParseException {
		// init
		consumer = null;
		heartbeatProcessor = null;
		heartbeatDirectory = hbDirectory;
		topicList = null;
		CONFIGTYPE_STRING = "ConsumerConfig";

		// init last  heartbeat time to now
		lastHeartbeatTime = (Long) (System.currentTimeMillis() / 1000);		

		// configuration/setup
		Properties configuration = null;
		try {
			configuration = convertJSONStringToProp(configString);
		} catch (ParseException e) {
			logger.error("ParseException converting configuration: " + e.toString());
			throw (e);
		}

		setup(configuration);		
	}

	/**
	 * The setup function for the Consumer class. Uses the provided Properties 
	 * configuration to configure the kafka producer client.
	 *
	 * @param configProperties
	 *            - A Properties containing the configuration
	 * @return Returns true if successful, false otherwise.
	 */	
	public boolean setup(Properties configProperties) {

		if (configProperties == null) {
			return(false);
		}

		try {
			// add any fixed configuration (like the serializer
			configProperties.put("key.deserializer",
					"org.apache.kafka.common.serialization.StringDeserializer");
			configProperties.put("value.deserializer",
					"org.apache.kafka.common.serialization.ByteArrayDeserializer");

			// create the consumer
			consumer = new KafkaConsumer<String, byte[]>(configProperties);

			// create the heartbeat processor, we need this to tell if a 
			// message is a heartbeat or not
			heartbeatProcessor = new Heartbeat();	
		} catch (Exception e) {
			logger.error("Exception configuring consumer: " + e.toString());
			return(false);
		}

		return(true);
	}

	/**
	 * Subscribes the kafka consumer client to the provided topic
	 *
	 * @param topic
	 *            - A String containing the topic to subscribe to.
	 */
	public void subscribe(String topic) {
		subscribe(Arrays.asList(topic));
	}

	/**
	 * Subscribes the kafka consumer client to the provided list of topics
	 *
	 * @param topics
	 *            - A Collection&lt;String&gt; containing the list of topics to
	 *            subscribe to.
	 */
	public void subscribe(Collection<String> topics) {
		consumer.subscribe(topics);

		// remember the topic list for handling heartbeats
		topicList = topics;
	}

	/**
	 * Polls the hazdev kafka broker cluster for data.
	 *
	 * @param timeout
	 *            - A long containing the time to wait while polling in
	 *            milliseconds. -1 indicates that the client should wait for an
	 *            arbitrarily long time
	 * @return Returns an ArrayList&lt;byte[]&gt; containing the data from the broker
	 *         cluster since the last time it was polled.
	 */
	public ArrayList<byte[]> poll(long timeout) {

		ArrayList<byte[]> data = new ArrayList<byte[]>();

		// Negative value means wait for an arbitrarily long time
		if (timeout < 0) {
			timeout = Long.MAX_VALUE;
		}

		// get any messages pending for our topic(s) from kafka
		ConsumerRecords<String, byte[]> records = null;
		try {
			records = consumer.poll(timeout);
		} catch (Exception e) { 
			logger.error("Error calling consumer.poll: " + e.toString());
			return (null);
		}

		// nullcheck
		if (records == null) {
			return(null);
		}

		// go though each message, adding it to the return ArrayList
		// removing heartbeat messages
		for (ConsumerRecord<String, byte[]> record : records) {
			// nullcheck
			if (record == null) {
				continue;
			}

			// convert to string to see if this is a heartbeat
			// note that if poll was called by pollString, we're
			// converting *twice*, I'm not sure how to check for heartbeats
			// more efficiently than this tho
			String recordString = new String(record.value());

			// don't add heartbeats to the data arraylist
			if (heartbeatProcessor.fromJSONString(recordString) == true) {
				handleHeartbeat(heartbeatProcessor);
			} else {
				data.add(record.value());
			}
		}

		return (data);
	}

	/**
	 * Polls the hazdev kafka broker cluster for messages.
	 *
	 * @param timeout
	 *            - A long containing the time to wait while polling in
	 *            milliseconds. -1 indicates that the client should wait for an
	 *            arbitrarily long time
	 * @return Returns an ArrayList&lt;String&gt; containing the messages from the
	 *         broker cluster since the last time it was polled.
	 */
	public ArrayList<String> pollString(long timeout) {

		ArrayList<String> messages = new ArrayList<String>();

		// get data
		ArrayList<byte[]> data = poll(timeout);

		// nullcheck
		if (data == null) {
			return(null);
		}

		// convert bytes to strings
		for (byte[] aData : data) {
			// nullcheck
			if (aData == null) {
				continue;
			}
			
			messages.add(new String(aData));
		}

		return (messages);
	}

	/**
	 * heartbeat handling function
	 *
	 * @param aHeartbeat
	 *            - A Heartbeat containing the heartbeat to handle
	 */	
	public void handleHeartbeat(Heartbeat aHeartbeat) {

		// is this a valid heartbeat
		if (aHeartbeat.isValid() == false) {
			return;
		}

		// is this heartbeat for one of the configured topics
		// we don't want to handle hearbeats for other topics
		if (!topicList.contains(aHeartbeat.getTopic())) {
			return;
		}

		// set the time the heartbeat was received in case our 
		// caller is monitoring this
		setLastHeartbeatTime(System.currentTimeMillis() / 1000);

		// write the heartbeat to disk (won't write if heartbeatDirectory is
		// null)
		try {
			aHeartbeat.writeToDisk(heartbeatDirectory);
		} catch (FileNotFoundException e) {
			logger.error("FileNotFoundException " + e.toString());
			return;
		} catch (UnsupportedEncodingException e) {
			logger.error("UnsupportedEncodingException " + e.toString());
			return;
		}  
	}

	/**
	 * this function gets the partition ids for a topic
	 *
	 * @param topic
	 *            - A string containing the topic to query
	 */	
	public ArrayList<String> getPartitions(String topic) {
		ArrayList<String> partitions = new ArrayList<String>();

		if (consumer == null) {
			return partitions;
		}

		try{
			List<PartitionInfo> partitionInfo = consumer.partitionsFor(topic);

			for (int i = 0; i < partitionInfo.size(); i++) {
				PartitionInfo aPartition = partitionInfo.get(i);
				partitions.add(String.valueOf(aPartition.partition()));
			}
		} catch (Exception e) {
		}

		return partitions;
	}

	/**
	 * @return the lastHeartbeatTime
	 */
	public Long getLastHeartbeatTime() {
		return lastHeartbeatTime;
	}

	/**
	 * @param lastHeartbeatTime
	 *            the lastHeartbeatTime to set
	 */
	public void setLastHeartbeatTime(Long lastHeartbeatTime) {
		this.lastHeartbeatTime = lastHeartbeatTime;
	}  
}

File: cpp/src/Consumer.cpp
#include <Consumer.h>
#include <rdkafkacpp.h>
#include <limits>
#include <cmath>
#include <algorithm>

// Define configuration type
#define CONFIGTYPE_STRING "ConsumerConfig"

namespace hazdevbroker {

Consumer::Consumer() {
	// init
	m_pConsumer = NULL;
	m_sConfigType = std::string(CONFIGTYPE_STRING);
	m_TopicList.clear();
	m_sHeartbeatDirectory = "";
	m_lLastHeartbeatTime = std::time(NULL);
}

Consumer::Consumer(rapidjson::Value &configJSON, // NOLINT
		rapidjson::Value &topicConfigJSON) { // NOLINT
	// init
	m_pConsumer = NULL;
	m_sConfigType = std::string(CONFIGTYPE_STRING);
	m_TopicList.clear();
	m_sHeartbeatDirectory = "";
	m_lLastHeartbeatTime = std::time(NULL);

	// setup using json
	setup(configJSON, topicConfigJSON);
}

Consumer::Consumer(std::string configString, std::string topicConfigString) {

	// init
	m_pConsumer = NULL;
	m_sConfigType = std::string(CONFIGTYPE_STRING);
	m_TopicList.clear();
	m_sHeartbeatDirectory = "";
	m_lLastHeartbeatTime = std::time(NULL);

	// setup using a string
	setup(configString, topicConfigString);
}

Consumer::Consumer(rapidjson::Value &configJSON, // NOLINT
		rapidjson::Value &topicConfigJSON, std::string hbDirectory) { // NOLINT
	// init
	m_pConsumer = NULL;
	m_sConfigType = std::string(CONFIGTYPE_STRING);
	m_TopicList.clear();
	m_sHeartbeatDirectory = hbDirectory;
	m_lLastHeartbeatTime = std::time(NULL);

	// setup using json
	setup(configJSON, topicConfigJSON);
}

Consumer::Consumer(std::string configString, std::string topicConfigString,
		std::string hbDirectory) {

	// init
	m_pConsumer = NULL;
	m_sConfigType = std::string(CONFIGTYPE_STRING);
	m_TopicList.clear();
	m_sHeartbeatDirectory = hbDirectory;
	m_lLastHeartbeatTime = std::time(NULL);

	// setup using a string
	setup(configString, topicConfigString);
}

Consumer::~Consumer() {
	// cleanup
	if (m_pConsumer != NULL) {
		m_pConsumer->close();
		delete (m_pConsumer);
	}
}

void Consumer::setup(rapidjson::Value &configJSON,
		rapidjson::Value &topicConfigJSON) {
	std::string errstr;

	// create config
	RdKafka::Conf *conf = convertJSONConfigToProp(configJSON, topicConfigJSON);

	// config created?
	if (conf == NULL) {
		m_pConsumer = NULL;
		return;
	}

	// create consumer
	m_pConsumer = RdKafka::KafkaConsumer::create(conf, errstr);

	// error check
	if (!m_pConsumer) {
		log("Error Creating consumer: " + errstr);
		m_pConsumer = NULL;
	}

	// cleanup
	delete conf;
}

void Consumer::setup(std::string configString, std::string topicConfigString) {
	std::string errstr;

	// create config
	RdKafka::Conf *conf = convertJSONStringToProp(configString,
			topicConfigString);

	// config created?
	if (conf == NULL) {
		m_pConsumer = NULL;
		return;
	}

	// create consumer
	m_pConsumer = RdKafka::KafkaConsumer::create(conf, errstr);

	// error check
	if (!m_pConsumer) {
		log("Error Creating consumer: " + errstr);
		m_pConsumer = NULL;
	}

	// cleanup
	delete conf;
}

bool Consumer::subscribe(std::string topic) {
	// nullcheck
	if (m_pConsumer == NULL) {
		return (false);
	}

	// convert topic string to vector
	std::vector <std::string> topics;
	topics.push_back(topic);

	// subscribe
	return (subscribe(topics));
}

bool Consumer::subscribe(std::vector<std::string> topics) {
	// nullcheck
	if (m_pConsumer == NULL) {
		return (false);
	}

	// subscribe to the topics
	RdKafka::ErrorCode err = m_pConsumer->subscribe(topics);

	// error check
	if (err) {
		log("Error subscribing to topic(s): " + RdKafka::err2str(err));
		return (false);
	}

	// remember the topic list for handling heartbeats
	for (auto topic : topics) {
		m_TopicList.push_back(topic);
	}

	return (true);
}

byte * Consumer::poll(int64_t timeout, size_t *datalength) {
	byte * data = NULL;

	// nullcheck
	if (m_pConsumer == NULL) {
		return (NULL);
	}

	// Negative value means wait for an arbitrarily long time
	if (timeout < 0) {
		timeout = std::numeric_limits<int64_t>::max();
	}

	// get any messages pending for our topic(s) from kafka
	RdKafka::Message *msg = m_pConsumer->consume(timeout);

	// we get something?
	if (msg->err() == RdKafka::ERR_NO_ERROR) {
		// get length
		*datalength = msg->len();

		// copy data
		data = new byte[*datalength];
		memcpy(data, static_cast<byte *>(msg->payload()), *datalength);

	} else if (msg->err() == RdKafka::ERR__TIMED_OUT) {
		// cleanup
		delete (msg);

		// got nothing
		*datalength = 0;
		return (NULL);
	} else if (msg->err() == RdKafka::ERR__PARTITION_EOF) {
		// cleanup
		delete (msg);

		// got nothing
		*datalength = 0;
		return (NULL);
	} else {
		// error
		log("Error polling broker cluster " + RdKafka::err2str(msg->err())
			+ ": " + std::string(msg->errstr()));

		// cleanup
		delete (msg);

		// got nothing
		*datalength = 0;
		return (NULL);
	}

	// cleanup
	delete (msg);

	// handle  heartbeats
	if ((data != NULL) && (*datalength != 0)) {
		// convert to string
		std::string stringData = "";
		try {
			// convert bytes to string
			stringData = std::string(reinterpret_cast<const char *>(data),
					*datalength);
		} catch (const std::exception &e) {
			// we can safely ignore this, if it can't convert, it's
			// definitely not a heartbeat
			return (data);
		}

		// check if this is a heartbeat
		if (m_HeartbeatProcessor.fromJSONString(stringData) == true) {
			// handle it
			handleHeartbeat(m_HeartbeatProcessor);

			// cleanup heartbeat
			delete (data);

			// we don't return anything for heartbeats
			*datalength = 0;
			return (NULL);
		} else {
			// done
			return (data);
		}
	}

	// done
	return (data);
}

std::string Consumer::pollString(int64_t timeout) {
	size_t length = 0;
	byte* byteMessage = NULL;

	// nullcheck
	if (m_pConsumer == NULL) {
		return ("");
	}

	// get data
	byteMessage = poll(timeout, &length);

	// did we get anything?
	if ((byteMessage == NULL) || (length == 0)) {
		return ("");
	}

	std::string stringMessage = "";
	try {
		// convert bytes to string
		stringMessage = std::string(reinterpret_cast<const char *>(byteMessage),
				length);
	} catch (const std::exception &e) {
		log("Exception converting bytes to string: " + std::string(e.what()));
		return ("");
	}

	// cleanup
	delete (byteMessage);

	// done
	return (stringMessage);
}

void Consumer::handleHeartbeat(hazdevbroker::Heartbeat aHeartbeat) {
	// is this a valid heartbeat
	if (aHeartbeat.isValid() == false) {
		return;
	}

	// is this heartbeat for one of the configured topics
	// we don't want to handle hearbeats for other topics
	if(std::find(m_TopicList.begin(), m_TopicList.end(), aHeartbeat.topic)
			== m_TopicList.end()) {
		return;
	}

	// set the time the heartbeat was recieved in case our
	// caller is monitoring this
	setLastHeartbeatTime(std::time(NULL));

	// write the heartbeat to disk (won't write if heartbeatDirectory is
	// empty)
	aHeartbeat.writeToDisk(m_sHeartbeatDirectory);
}

int64_t Consumer::getLastHeartbeatTime() {
	return(m_lLastHeartbeatTime);
}

void Consumer::setLastHeartbeatTime(int64_t lastHeartbeatTime) {
	m_lLastHeartbeatTime = lastHeartbeatTime;
}

std::string Consumer::getHeartbeatDirectory() {
	return(m_sHeartbeatDirectory);
}

void Consumer::setHeartbeatDirectory(std::string heartbeatDirectory) {
	m_sHeartbeatDirectory = heartbeatDirectory;
}
}  // namespace hazdevbroker
